packages/router/src/helpers.rs

```diff
@@ -0,0 +1,240 @@
+use std::{collections::BTreeMap, sync::Arc};
+
+use dioxus::prelude::*;
+use log::error;
+use urlencoding::encode;
+
+use crate::{
+    contexts::RouterContext,
+    navigation::{NamedNavigationSegment, Query},
+};
+
+/// A private "hook" to subscribe to the router.
+///
+/// Used to reduce redundancy within other components/hooks. Safe to call multiple times for a
+/// single component, but not recommended. Multiple subscriptions will be discarded.
+///
+/// # Return values
+/// - [`None`], when the current component isn't a descendant of a [`Router`].
+/// - Otherwise [`Some`].
+///
+/// [`Router`]: crate::components::router
+#[must_use]
+#[allow(clippy::mut_from_ref)]
+pub(crate) fn sub_to_router(cx: &ScopeState) -> &mut Option<RouterContext> {
```

Comment:
I'm not really sure if it is a hook, since it isn't exposed to users. That's why I put the quotation marks around `"hook"`.

Reply:
well it makes a call to `use_hook`, so it has to be a hook (or a component)

this is important, because `use_hook` (and hooks, by extension) [have certain "rules" that the caller must follow](https://dioxuslabs.com/nightly/guide/interactivity/hooks.html#rules-of-hooks) for correctness, even if it's just an internal API. for this reason, i would reccomend naming it as a hook. then:

- it's more clear what it does (hooks initialize some stuff on the first render)
- it's more clear that hook rules must be followed
- hook conventions are followed

Reply:
Yeah, good points. Changed it to `use_router_subscription`.